### PR TITLE
Tests for C++ Autograd API

### PR DESCRIPTION
**Pull request by malvika2147**
### All the PRs in this stack have been moved to #23572, abandoning this PR

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23533 Tests for C++ Autograd API**
* #23184 API for AutogradContext
* #23020 Support custom autograd functions in C++

Summary: Tests for the custom autograd function API based on test_autograd.py. Currently only the tests for the basic functionality have been added. More tests will be added later.

### CODE REVIEW

File: test/cpp/api/autograd.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+#include <torch/autograd.h>
+
+#include <torch/utils.h>
+#include <test/cpp/api/support.h>
+
+using namespace torch;
+
+#define ASSERT_VARIABLE_EQ(a,b) ASSERT_TRUE(torch::allclose((a),(b)))
+
+std::string graph_desc(std::shared_ptr<autograd::Node> node) {
```
Comment (ezyang) [inline test/cpp/api/autograd.cpp]: Since this is testing code, it doesn't really matter, but for future reference: if you need to cat strings together in C++, best to use `std::stringstream`.

File: test/cpp/api/autograd.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  for(auto& edge : next_edges) {
+    result += graph_desc(edge.function);
+  }
+  return result+")";
+}
+
+TEST(CustomAutogradTest, CustomFunction) {
+  struct MyFunction : public torch::autograd::Function<MyFunction> {
+    static variable_list forward(AutogradContext *ctx, Variable var1, int mul, Variable var2) {
+      ctx->saved_data["mul"] = mul;
```
Comment (ezyang) [inline test/cpp/api/autograd.cpp]: I guess I take my comment back about a getter/setter API; this is pretty natural looking.

File: test/cpp/api/autograd.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+      ctx->save_for_backward({var1, var2});
```
Comment (ezyang) [inline test/cpp/api/autograd.cpp]: Maybe we should have made this a variadic C++ function. Then it would match Python API even better :)

File: test/cpp/api/autograd.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      return {var1 + mul*var2 + var1*var2};
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
+      int mul = ctx->saved_data["mul"].toInt();
+      auto saved = ctx->get_saved_variables();
+      auto var1 = saved[0];
+      auto var2 = saved[1];
```
Comment (ezyang) [inline test/cpp/api/autograd.cpp]: Bloop, it's too bad we can't use `std::tie` here (we'd need a tuple, not a vector >:( )

File: test/cpp/api/autograd.cpp [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      variable_list output = {grad_output[0] + grad_output[0]*var2, Variable(), grad_output[0] * mul + grad_output[0] * var1};
+      return output;
+    }
+  };
+
+  Variable x = torch::randn({5,5}, torch::requires_grad());
+  Variable y = torch::randn({5,5}, torch::requires_grad());
+  auto res = MyFunction::apply(x,2,y)[0];
```
Comment (ezyang) [inline test/cpp/api/autograd.cpp]: Hmm, I guess we could probably do some shenanigans to make single output functions nicer (extra template parameter to Function which specifies the return type of apply), but that'd be some extra C++ work.

Comment (ezyang) [reply test/cpp/api/autograd.cpp]: Maybe we can file an enhancement request for this and see if someone wants to pick it up.

**Review by ezyang — APPROVED** [review]

File: test/cpp/api/autograd.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+  Variable x = torch::randn({5,5});
+  Variable y = torch::randn({5,5}, torch::requires_grad());
+
+  auto out = Inplace::apply(x,y);
+  auto &q = out[0];
+  ASSERT_TRUE(torch::equal(q, x));
+  ASSERT_TRUE(q.requires_grad());
+  q.sum().backward();
+  ASSERT_VARIABLE_EQ(y.grad(), torch::ones({5,5}));
+}
```
Comment (yf225) [inline test/cpp/api/autograd.cpp]: Should we also add a test case for returning a null gradient in backward(), similar to test_hook_none in test_autograd.py?

File: test/cpp/api/autograd.cpp (hunk 4 above)
Comment (yf225) [reply test/cpp/api/autograd.cpp]: Yes it would be awesome to be able to use std::tie here to further match the Python API, since the Python API can do this in one line :)

**Comment by malvika2147** [comment]
squashed the entire stack into a new PR

File: test/cpp/api/autograd.cpp (hunk 6 above)
Comment (malvika2147) [reply test/cpp/api/autograd.cpp]: The first test (CustomFunction) has a backward function that returns an empty Variable, which is treated the same way as "null" in the python code. Does that count? 
Also, I'll be adding all the other api tests from test_autograd.py in future PR.

Comment (yf225) [reply test/cpp/api/autograd.cpp]: I see, I think since `variable_list` only accepts Variables (and can't convert `null`), using `Variable()` is a good idea.
